### PR TITLE
3+D Parameter Sweep Plots

### PR DESCRIPTION
**Pull request by BryanRumsey**

### CODE REVIEW

**Comment by briandrawert** [comment]
Screenshot from 2021-07-29 11-39-30
Extra Quotes around "Review Model" on parameter sweep page.

**Review by briandrawert — APPROVED** [review]

File: stochss/handlers/util/parameter_sweep_2d.py [hunk 1]
```diff
@@ -140,32 +141,74 @@ def get_plotly_traces(self, keys):
...
+                red_results = [map_result[0] for map_result in map_results]
+            data.append(red_results)
+        data = numpy.array(data)
+
+        trace_list = [plotly.graph_objs.Heatmap(z=data, x=params[0]['range'],
+                                                y=params[1]['range'])]
+
+        title = f"<b>Parameter Sweep - Variable: {species}</b>"
+        layout = plotly.graph_objs.Layout(title=dict(text=title, x=0.5),
+                                          xaxis=dict(title=f"<b>{params[0]['name']}</b>"),
```
Comment (jtcooper10) [inline stochss/handlers/util/parameter_sweep_2d.py:208]: This is a XSS vulnerability, and allows for arbitrary code execution on published models.

To replicate: set your species name to a JavaScript code segment, like `<script>window.alert("ping!")</script>` and run a parameter sweep simulation. It's validated on the front end, but not on the server.

image

File: stochss/handlers/util/parameter_sweep_1d.py [hunk 2]
```diff
@@ -143,26 +144,69 @@ def get_plotly_traces(self, keys):
...
-            results = self.results[keys[0]][keys[1]]
+            data = [[map_result[0], 0] for map_result in map_results]
+            visible = False
+        data = numpy.array(data)
+
+        error_y = dict(type="data", array=data[:, 1], visible=visible)
+        trace_list = [plotly.graph_objs.Scatter(x=param['range'],
+                                                y=data[:, 0], error_y=error_y)]
+
+        title = f"<b>Parameter Sweep - Variable: {species}</b>"
```
Comment (jtcooper10) [inline stochss/handlers/util/parameter_sweep_1d.py:203]: Same thing here: inline species name getting returned to the client is a XSS vulnerability

**Review by jtcooper10 — CHANGES_REQUESTED** [review]
Everything works except for what I left comments about. Tested in Edge on Windows 11.

**Comment by briandrawert** [comment]
`overflow-x: scroll;`  on the parameter sliders creates scrollbars when there should be none

**Comment by briandrawert** [comment]
When a checkbox is clicked on the sweep tab, 'disable' all other "fixed" sliders if their fixed checkbox is not selected.

File: client/templates/includes/sweepParameterRange.pug [hunk 3]
```diff
@@ -15,7 +15,7 @@ div
 
     div.inline(style="max-width: 100%")
 
-      div.horizontal-scroll(data-hook=this.model.elementID + "-current-value")=this.model.range[0]
+      div.horizontal-scroll(data-hook=this.model.elementID + "-current-value" style="overflow-x: none;")=this.model.range[0]
```
Comment (briandrawert) [inline client/templates/includes/sweepParameterRange.pug:18]: none => visible

**Review by briandrawert — APPROVED** [review]

**Pull request merged**